proto/private/utils/direct_source_infos.bzl

```diff
@@ -0,0 +1,72 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Contains the implementation for `proto_common.direct_source_infos`."""
+
+ProtoFileInfo = provider(
+    doc = "Encapsulates information about a single `.proto` file.",
+    fields = {
+        "file": "The source file.",
+        "import_path": "The import path of the `.proto` file.",
+    },
+)
+
+# TODO(yannic): Remove this suppression when
+# https://github.com/bazelbuild/buildtools/issues/774 has been fixed.
+# buildifier: disable=function-docstring-args
+def direct_source_infos(proto_info, *, provided_sources = []):
```

Comment:
Why isn't this a member of `ProtoInfo`? Sounds like a reasonable functionality there. (I might be missing something obvious here, since I haven't been following this pull request very closely)

Reply:
I thought about making import paths of direct sources part of `ProtoInfo` before but never came to a conclusion.

`ProtoInfo` already is "complicated" (:cough: original sources) and I'd like to avoid adding more half-baked stuff to it. Can you file a feature request?